### ISSUE TITLE
Add upgrade notice for breaking changes in readme.

### ISSUE THREAD

**Issue by aristath** [themeum/kirki#1327]

**Comment by carasmo** [comment]
Not exactly sure what this is. So is it upgrades that will break things from previous versions? If so, how will the non-dev theme user know whether or not to upgrade?

**Comment by aristath** [comment]
I've paid a lot of attention not to introduce any breaking changes, and in case I find any I'm trying to fix them.
BUT some things are inevitable.

So far the only breaking change I have found is with `background` controls.
For example, in pre-1.0 versions ( more than 2 years ago) we had a "background" field-type.
That wasn't exactly a field-type, it was simply adding multiple separate controls, grouped them together in the UI, and then saved them separately in the db.
If you haven't seen this before it's because it was awful and I didn't even add it in the docs as an available field-type.
However, some of the early themes that were built using Kirki were using it (I don't know if any of them are around anymore)
In version 3.0 I added a proper `background` control that saves the values as an array (background-color, background-image, background-repeat etc), has `output` and `transport` working fine and so on.

It may have been something that was added 2 years ago, but I kept the implementation around until v2.3.7 (see [this file](https://github.com/aristath/kirki/blob/2.3.7/includes/class-kirki-explode-background-field.php)) to accommodate these themes, and whenever someone mentioned that they're using the background field in their themes I prompted them to change to separate controls instead because it would be deprecated at some point.
Wether they followed my advice or not is a mystery to me...

But the point is that there's no way to provide backwards-compatibility for this, unless I do things that I would classify as bad. 

Example:
Say you created a background control with an id (setting) `background_setting`.

NEW DB structure:
```php
$theme_mods = array(
	'background_setting' => array(
		'background-color'      => '#ffffff',
		'background-image'      => 'url-to-the-image',
		'background-repeat'     => 'repeat-all',
		'background-position'   => 'left top',
		'background-size'       => 'cover',
		'background-attachment' => 'fixed',
	),
);
```
OLD DB structure: (it was appending the property to the setting-id)
```php
$theme_mods = array(
	'background_setting_color'      => '#ffffff',
	'background_setting_image'      => 'url-to-the-image',
	'background_setting_repeat'     => 'repeat-all',
	'background_setting_position'   => 'left top',
	'background_setting_size'       => 'cover',
	'background_setting_attachment' => 'fixed',
);
```

A possible solution: Save separate values in the db (in addition to the one saved as an array).
```php
$theme_mods = array(
	'background_setting' => array(
		'background-color'      => '#ffffff',
		'background-image'      => 'url-to-the-image',
		'background-repeat'     => 'repeat-all',
		'background-position'   => 'left top',
		'background-size'       => 'cover',
		'background-attachment' => 'fixed',
	),
	'background_setting_color'      => '#ffffff',
	'background_setting_image'      => 'url-to-the-image',
	'background_setting_repeat'     => 'repeat-all',
	'background_setting_position'   => 'left top',
	'background_setting_size'       => 'cover',
	'background_setting_attachment' => 'fixed',
);
```

That last one would provide compatibility with older versions, but at the same time will mess things up in case you've got a separate field with id=`background_setting_color` for instance.

**Comment by aristath** [comment]
@carasmo btw if you have any ideas on how to handle the above do tell 😄

**Comment by carasmo** [comment]
The last idea is great -- we are used to not using reserved classes, functions, and slugs. CMB2 docs want us to prefix our settings, and we should do the same here too. 

```
$theme_mods = array(
	'background_setting' => array(
		'background-color'      => '#ffffff',
		'background-image'      => 'url-to-the-image',
		'background-repeat'     => 'repeat-all',
		'background-position'   => 'left top',
		'background-size'       => 'cover',
		'background-attachment' => 'fixed',
	),
	'background_setting_color'      => '#ffffff',
	'background_setting_image'      => 'url-to-the-image',
	'background_setting_repeat'     => 'repeat-all',
	'background_setting_position'   => 'left top',
	'background_setting_size'       => 'cover',
	'background_setting_attachment' => 'fixed',
);
```

**Comment by aristath** [comment]
I still hate the idea of adding useless data in the db - and potentially breaking things like not allowing you to add a `background_setting_image` field if you have a `background_setting` field... 
It makes no sense.
Both methods have their pros and cons.

If I don't add this fallback there's the potential to break themes that are old and haven't been updated for a looooong time. Which is bad, but at the same time understandable. I mean if you neglect any piece of software for 2 years and don't update it to keep up with its dependencies (and by "dependencies" I also mean WP Core), then of course at some point it will break if you only update half the things in it. You can't expect a 4-year-old plugin to play nicely with a new theme and vice-versa. If you're using an old theme you're taking a big risk by updating anything on your site.

On the other hand if I do add this fallback, then I'll be adding things that 99% of people won't use just so that a handful of cases don't break. And at the same time I'll be limiting the ability to name settings however you want - which of course people will never know because they rarely read the docs - and then they'll try to add a `background_setting_repeat` field when they already have a `background_setting` and of course things will break. And it makes no sense that it breaks, it shouldn't!

This is a difficult decision that is 100% due to poor past choices.
So the real question here is this:

Do 2 wrongs make 1 right? In my experience that is never the case - as it shouldn't. 
**Does a wrong choice in the past justify another wrong choice now? It may be best for the present, but what about the future?**

**Comment by aristath** [comment]
On the bright side, if you were using a background control with the `output` and NOT by writing your own custom code to get the values and calculate the CSS, then your code will still work.
So the breaking cases are limited to themes that used kirki a long time ago and have a custom implementation to get background properties.

**Comment by carasmo** [comment]
Gotta go with your gut. WooCommerce just made major updates, very painful for many who don't bother to read, your users have less of a chance of that, except that a popular tut source has background documented (https://code.tutsplus.com/tutorials/toolbox-of-the-smart-wordpress-developer-kirki--cms-23739). But, as you said, if that tut doesn't keep up, it's not your fault. However the brunt of the pain will probably be felt in GitHub as I barely see any replies on their comments. I added one just now.

**Comment by aristath** [comment]
Actually the tutorial there is ok... it doesn't go in depth and just mentions that there's a `background` control.
That is still valid, and the syntax used to add the field is still identical.
The problem lies only if someone was trying to get individual values (for example just the background-color) using get_theme_mod.
But the existence of that tutorial may have led more people to use that control - even if it's not documented anywhere in the official docs. 😞 

I think I'll probably send a newsletter to the people that have subscribed in kirki.org, then email individually developers that have contacted me over time regarding background controls, and give everyone a week or two before the final release. I can't think of anything more that can be done if we want to improve things.
As for the upgrade notice, it will probably be something simple like 

> This is a major update, please keep a backup of your files and database before updating.

It's something that everyone should do anyway before performing updates on their site, but just to be on the safe side adding a notice like that won't hurt. 😄

**Comment by aristath** [comment]
Just an update here.
Tweeted about it, send a newsletter and let everyone know release will be around June 15 - so in about a couple of weeks.
Yesterday I released version 2.3.8 which does nothing except add the ability to use update notices (this is a really weird one... the **installed** version of the plugin must have a `in_plugin_update_message-{plugin_basename}` filter in order for notices to be displayed for future releases. So this could not be added in 3.0, it had to already be present on installed sites).
